manticore/core/smtlib/solver.py

```diff
@@ -143,207 +167,162 @@ def minmax(self, constraints, x, iters=10000):
-                self._stop_proc()
-            # self._proc.stdin.writelines(('(exit)\n',))
-            # self._proc.wait()
-        except Exception as e:
-            logger.error(str(e))
-
-    def _reset(self, constraints: Optional[str] = None) -> None:
-        """Auxiliary method to reset the smtlib external solver to initial defaults"""
-        if self._proc is None:
-            self._start_proc()
-        else:
-            if self.support_reset:
-                self._send("(reset)")
-
-                for cfg in self._init:
-                    self._send(cfg)
-            else:
-                self._stop_proc()
-                self._start_proc()
-        if constraints is not None:
-            self._send(constraints)
-
-    def _send(self, cmd: str) -> None:
+            self._proc.stdin.close()
+            self._proc.stdout.close()
+            # Kill the process
+            self._proc.kill()
+            # Wait for termination, to avoid zombies.
+            self._proc.wait()
+        self._proc = None
+
+    def __readline_and_count(self):
+        assert self._proc
+        assert self._proc.stdout
+        buf = self._proc.stdout.readline()  # No timeout enforced here
+        # lparen, rparen = buf.count("("), buf.count(")")
+        lparen, rparen = map(sum, zip(*((c == "(", c == ")") for c in buf)))
+        return buf, lparen, rparen
+
+    def send(self, cmd: str) -> None:
         """
         Send a string to the solver.
 
         :param cmd: a SMTLIBv2 command (ex. (check-sat))
         """
-        # logger.debug('>%s', cmd)
-        try:
-            if self._proc.stdout:
-                self._proc.stdout.flush()
-            else:
-                raise SolverError("Could not flush stdout: file descriptor is None")
-            if self._proc.stdin:
-                self._proc.stdin.write(f"{cmd}\n")
-            else:
-                raise SolverError("Could not write to stdin: file descriptor is None")
-        except IOError as e:
-            raise SolverError(str(e))
-
-    def _recv(self) -> str:
-        """Reads the response from the solver"""
+        assert self._proc
+        assert self._proc.stdout
+        assert self._proc.stdin
+        if self._debug:
+            logger.debug(">%s", cmd)
+            print(">", cmd)
+        self._proc.stdout.flush()
+        self._proc.stdin.write(f"{cmd}\n")
+
+    def recv(self) -> str:
+        """Reads the response from the smtlib solver"""
         buf, left, right = self.__readline_and_count()
         bufl = [buf]
-
         while left != right:
             buf, l, r = self.__readline_and_count()
             bufl.append(buf)
             left += l
             right += r
-            if "(error" in bufl[0]:
-                raise SolverException(f"Error in smtlib: {bufl[0]}")
 
         buf = "".join(bufl).strip()
-        if "(error" in bufl[0]:
-            raise SolverException(f"Error in smtlib: {bufl[0]}")
```

Comment:
Are we expecting this error to not be able to occur anymore? It doesn't look like we check for either this case or the similar case on line 330. 

Also: why do we need both `SolverError` and `SolverException`?

Reply:
SolverError and SolverException? I do not think we need both. Let's choose one. ??

Ah!!  Good catch. Tried to ignore it in my refactor it seems. The problem there was that the thing after the "error" keyword could be a text containing unbalanced parenthesis. Need to think more about it.

Why tests pass and this does not occur normally?
We control the stmlib and this `error` only happen when we change something on the wrong side in the smtlib generation. And the smtlib parser of the solver complains.

Reply:
Added the _syntax error_ check elsewhere. Though only enabled when debugging.